addons/sourcemod/scripting/ConnectAnnounce.sp

```diff
@@ -422,164 +458,327 @@ public Action Command_Announce(int client, int args)
+
+		g_DatabaseState = DatabaseState_Connecting;
+		g_iConnectLock = g_iSequence++;
+		Database.Connect(GotDatabase, DATABASE_NAME, g_iConnectLock);
+	}
+
+	return false;
 }
 
-stock void OnSQLConnected(Handle hParent, Handle hChild, const char[] err, any data)
+public void GotDatabase(Database db, const char[] error, any data)
 {
-	if (hChild == null)
+	// See if the connection is valid.
+	if (db == null)
 	{
-		SetFailState("Failed to connect to database \"%s\". (%s)", DATABASE_NAME, err);
+		LogError("Connecting to database \"%s\" failed: %s", DATABASE_NAME, error);
 		return;
 	}
 
-	char sDriver[16];
-	g_hDatabase = CloneHandle(hChild);
-	SQL_GetDriverIdent(hParent, sDriver, sizeof(sDriver));
+	LogMessage("Connected to database.");
+
+	//PrintToServer("GotDatabase(data: %d, lock: %d, g_h: %d, db: %d)", data, g_iConnectLock, g_hDatabase, db);
+
+	// If this happens to be an old connection request, ignore it.
+	if (data != g_iConnectLock || (g_hDatabase != null && g_DatabaseState == DatabaseState_Connected))
+	{
+		if (db)
+			delete db;
+		return;
+	}
 
-	SQL_LockDatabase(g_hDatabase);
+	char sDriver[16];
+	SQL_GetDriverIdent(g_hDatabase, sDriver, sizeof(sDriver));
 
 	if (!strncmp(sDriver, "my", 2, false))
 		g_bSQLite = false;
 	else
 		g_bSQLite = true;
 
-	SQLSetNames(INVALID_HANDLE);
+	g_iConnectLock = 0;
+	g_DatabaseState = DatabaseState_Connected;
+	g_hDatabase = db;
+
+	DB_SetNames(db);
+	DB_CreateTable(db);
+}
 
-	SQLTableCreation_Join(INVALID_HANDLE);
+stock bool DB_Conn_Lost(DBResultSet db)
+{
+	if (db == null)
+	{
+		if (g_hDatabase != null)
+		{
+			LogError("Lost connection to DB. Reconnect after delay.");
+			delete g_hDatabase;
+			g_hDatabase = null;
+		}
 
-	SQL_UnlockDatabase(g_hDatabase);
+		if (g_DatabaseState != DatabaseState_Wait && g_DatabaseState != DatabaseState_Connecting)
+		{
+			g_DatabaseState = DatabaseState_Wait;
+			CreateTimer(RetryTime, Timer_DBReconnect, _, TIMER_FLAG_NO_MAPCHANGE);
+		}
+
+		return true;
+	}
+
+	return false;
 }
 
-stock Action SQLSetNames(Handle timer)
+stock void DB_Reconnect()
 {
-	if (!g_bSQLite)
-		SQL_TQuery(g_hDatabase, OnSqlSetNames, "SET NAMES \"UTF8MB4\"");
-	return Plugin_Stop;
+	g_DatabaseState = DatabaseState_Disconnected;
+	DB_Connect();
 }
 
-stock void OnSqlSetNames(Handle hParent, Handle hChild, const char[] err, any data)
+stock void DB_SetNames(Database db)
 {
-	if (hChild == null)
+	static int retries = 0;
+	char sQuery[MAX_SQL_QUERY_LENGTH];
+	Format(sQuery, sizeof(sQuery), "SET NAMES \"UTF8MB4\"");
+
+	if (DB_Connect())
 	{
-		SetFailState("Database error while setting names as utf8. (%s)", err);
-		return;
+		db.Query(Query_ErrorCheck, sQuery);
+	}
+	else
+	{
+		if (retries < g_cvQueryRetry.IntValue)
+		{
+			PrintToServer("[ConnectAnnounce] Failed to connect to database, retrying... (%d/%d)", retries, g_cvQueryRetry.IntValue);
+			PrintToServer("[ConnectAnnounce] Query: %s", sQuery);
+			DB_SetNames(db);
```

Comment:
dont call it directly like this because this will spam the database, instead add a timer of like 1.2 seconds * retries